Mathlib/RingTheory/KrullDimension.lean

```diff
@@ -0,0 +1,299 @@
+-/
+theorem le_of_quot (R : Type _) [CommRing R] (I : Ideal R) :
+    ringKrullDim (R ⧸ I) ≤ ringKrullDim R :=
+  le_of_surj _ _ (Ideal.Quotient.mk I) Ideal.Quotient.mk_surjective
+
+/--
+If `R` and `S` are isomorphic, then `ringKrullDim R = ringKrullDim S`.
+-/
+theorem eq_of_ringEquiv (R S : Type _) [CommRing R] [CommRing S] (e : R ≃+* S) :
+    ringKrullDim R = ringKrullDim S :=
+  le_antisymm (le_of_surj S R (RingEquiv.symm e) (EquivLike.surjective (RingEquiv.symm e)))
+    (le_of_surj R S e (EquivLike.surjective e))
+
+section field
+
+instance primeSpectrum_unique_of_field (F : Type _) [Field F] : Unique (PrimeSpectrum F) where
+  default := ⟨⊥, Ideal.bot_prime⟩
+  uniq := fun p ↦ PrimeSpectrum.ext _ _ $ Ideal.ext $ fun x ↦ by
+    refine ⟨fun h ↦ ?_, fun h ↦ h.symm ▸ Submodule.zero_mem _⟩
+    rwa [p.asIdeal.eq_bot_of_prime] at h
+
+instance finiteDimensionalType_of_field (F : Type _) [Field F] :
+  FiniteDimensionalOrder (PrimeSpectrum F) := inferInstance
+
+lemma eq_zero_of_Field (F : Type _) [Field F] : ringKrullDim F = 0 :=
+  krullDim.eq_zero_of_unique
+
+end field
+
+section dimension_0
+
+variable {R : Type _} [CommRing R]
+
+lemma _root_.Ideal.IsPrime.isMaximal_of_dim_zero
+    {I : Ideal R} (h : I.IsPrime) (dim0 : ringKrullDim R = 0) :
+    I.IsMaximal := by
+  obtain ⟨J, hJ1, hJ2⟩ := I.exists_le_maximal h.1
+  by_cases hIJ : I = J
+  · subst hIJ; exact hJ1
+  let x : LTSeries (PrimeSpectrum R) :=
+  { length := 1
+    toFun := ![⟨I, h⟩, ⟨J, hJ1.isPrime⟩]
+    step := by
+      intro i
+      fin_cases i
+      simp only [Fin.zero_eta, Fin.castSucc_zero, Matrix.cons_val_zero, Fin.succ_zero_eq_one,
+        Matrix.cons_val_one, Matrix.head_cons, ← PrimeSpectrum.asIdeal_lt_asIdeal]
+      refine ⟨hJ2, fun r ↦ ?_⟩
+      exact hIJ <| Eq.symm <| hJ1.eq_of_le h.1 r }
+  have : 1 ≤ ringKrullDim R :=
+    le_iSup (α := WithBot (WithTop ℕ)) (f := fun x : LTSeries _ ↦ x.length) x
+  rw [dim0] at this
+  norm_num at this
+
+lemma _root_.Ideal.IsPrime.isMinimal_of_dim_zero
+    {I : Ideal R} (h : I.IsPrime) (dim0 : ringKrullDim R = 0) :
+    I ∈ minimalPrimes R := by
+  obtain ⟨p, hp, le⟩ := Ideal.exists_minimalPrimes_le (I := ⊥) (J := I) bot_le
+  by_cases hIp : I = p
+  · subst hIp; exact hp
+  let x : LTSeries (PrimeSpectrum R) :=
+  { length := 1
+    toFun := ![⟨p, hp.1.1⟩, ⟨I, h⟩]
+    step := by
+      intro i
+      fin_cases i
+      simp only [Fin.zero_eta, Fin.castSucc_zero, Matrix.cons_val_zero, Fin.succ_zero_eq_one,
+        Matrix.cons_val_one, Matrix.head_cons, ← PrimeSpectrum.asIdeal_lt_asIdeal]
+      refine ⟨le, ?_⟩
+      contrapose! hIp
+      exact le_antisymm hIp le }
+  have : 1 ≤ ringKrullDim R :=
+    le_iSup (α := WithBot (WithTop ℕ)) (f := fun x : LTSeries _ ↦ x.length) x
+  rw [dim0] at this
+  norm_num at this
+
+end dimension_0
+
+section artinian_and_noetherian
+
+variable (R : Type _) [CommRing R] [Nontrivial R]
+
+lemma eq_zero_of_isArtinianRing [IsArtinianRing R] : ringKrullDim R = 0 := by
```

Comment:
Can you extract a TFAE lemma with the following:
+ `ringKrullDim R = 0`
+ for any two primes P ≤ Q we have P = Q
+ all primes are maximal
+ all primes are minimal primes
+ (maybe?) all maximal ideas are minimal primes

Reply:
Hmm, maybe not. I think the principle should be that we should give constructors of ringKrullDim = 0 requiring weakest possible conditions, but state strongest possible consequences under the ringKrullDim = 0 assumption. Under this principle, we should give constructors from:
+ all maximal ideals are minimal primes
+ all minimal primes are maximal among the primes
(both are weaker than "for any two primes P ≤ Q we have P = Q")

Given ringKrullDim = 0, the strongest consequences are:
+ for any two primes P ≤ Q we have P = Q
+ for any ideal, being prime is equivalent to being maximal, and equivalent to being a minimal prime

Reply:
Oh apparently it's also equivalent to the prime spectrum being a [T1Space](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Separation.html#T1Space) (all primes are maximal so all points are closed), and we have [discrete_of_t1_of_finite](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Separation.html#discrete_of_t1_of_finite)

Reply:
Your advice is great! We can use this to prove every 0-dimensional ring is isomorphic to product of all its prime localizations. Thanks!